### PR TITLE
Fix memory leaks in linux heap flow

### PR DESCRIPTION
**Pull request by giridharprasath**
# SQUASH ME

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [X] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Fix memory leak in  `bin_elf` and fix a leak in `dmht`
Add support for `aarch64` tcache parsing
...

**Test plan**

Tested on ARM64 build

...

**Closing issues**

...

### CODE REVIEW

File: librz/bin/bin.c [hunk 1]
```diff
@@ -200,13 +200,15 @@ RZ_API void rz_bin_reloc_free(RZ_NULLABLE RzBinReloc *reloc) {
...
 	/**
 	 * TODO: leak in bin_elf, but it will cause double free in bin_pe if free here,
 	 * Because in the bin_elf implementation RzBinObject->imports and RzBinObject->relocs->imports
 	 * are two pieces of data, but they are linked to each other in bin_pe
 	 */
-	//	rz_bin_import_free(reloc->import);
-	//	rz_bin_symbol_free(reloc->symbol);
+	rz_bin_import_free(reloc->import);
+	rz_bin_symbol_free(reloc->symbol);
+#endif
```
Comment (wargio) [inline librz/bin/bin.c]: this is wrong. the issue is in the PE code which generates the double free.

Comment (giridharprasath) [reply librz/bin/bin.c]: Yes, I am looking into that.

**Comment by giridharprasath** [comment]
@wargio @imbillow

Please review(one testcase failure) and lmk if we have to change this testcase output?

```
[XX] db/formats/pe/imports_tinyW7 PE: resolve reloc name by ordinal from sdb
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'ir
ii
' bins/pe/imports_tinyW7.exe
-- stdout
--- expected
+++ actual
@@ -4,5 +4,5 @@
 0x800004f4 0x00000234 SET_32 msvcrt_Ordinal_1268
 nth  vaddr      bind type lib      name         
 ------------------------------------------------
-284  ---------- NONE FUNC kernel32 FindAtomW
+284  0x00401048 NONE FUNC kernel32 Ordinal_284
 1268 0x00401034 NONE FUNC msvcrt   Ordinal_1268
```

**Comment by b1llow** [comment]
@giridharprasath  There's something wrong with this test.

```
[XX] db/formats/pe/imports_tinyW7 PE: resolve reloc name by ordinal from sdb
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'ir
ii
' bins/pe/imports_tinyW7.exe
-- stdout
--- expected
+++ actual
@@ -4,5 +4,5 @@
 0x800004f4 0x00000234 SET_32 msvcrt_Ordinal_1268
 nth  vaddr      bind type lib      name         
 ------------------------------------------------
-284  ---------- NONE FUNC kernel32 FindAtomW
+284  0x00401048 NONE FUNC kernel32 Ordinal_284
 1268 0x00401034 NONE FUNC msvcrt   Ordinal_1268
```

File: librz/bin/p/bin_mach0.c [hunk 2]
```diff
@@ -408,9 +408,9 @@ static RzPVector /*<RzBinReloc *>*/ *relocs(RzBinFile *bf) {
 				free(ptr);
 				break;
 			}
-			ptr->import = imp;
+			ptr->import = rz_bin_import_clone(imp);
 		} else if (reloc->ord >= 0 && reloc->ord < rz_pvector_len(&bin->imports_by_ord)) {
-			ptr->import = rz_pvector_at(&bin->imports_by_ord, reloc->ord);
+			ptr->import = rz_bin_import_clone(rz_pvector_at(&bin->imports_by_ord, reloc->ord));
```
Comment (wargio) [inline librz/bin/p/bin_mach0.c]: even tho this is likely ok, do first an assignment to `RzBinImport *imp;` and then clone to `ptr->import`

File: librz/bin/bin.c [hunk 3]
```diff
@@ -205,8 +208,8 @@ RZ_API void rz_bin_reloc_free(RZ_NULLABLE RzBinReloc *reloc) {
 	 * Because in the bin_elf implementation RzBinObject->imports and RzBinObject->relocs->imports
 	 * are two pieces of data, but they are linked to each other in bin_pe
 	 */
```
Comment (wargio) [inline librz/bin/bin.c]: since this is fixed, you can remove this comment.

**Comment by XVilka** [comment]
Interesting, `FindAtomW` is 283 ordinal, not 284, see `librz/bin/d/dll/kernel32.sdb.txt`. So the previous output was a bug it seems?

**Comment by giridharprasath** [comment]
Ghidra's imports:
[
ScreenShot_2024-04-18_at_09:55:57-PM
](url)
ScreenShot_2024-04-18_at_09:49:34-PM

**Comment by giridharprasath** [comment]
Debug binary with `dmht` command:
Before fix: 
ScreenShot_2024-04-18_at_10:04:48-PM

After fix:
ScreenShot_2024-04-18_at_10:05:05-PM

File: librz/core/linux_heap_glibc.c [hunk 4]
```diff
@@ -849,11 +856,17 @@ static bool GH(parse_tcache_from_addr)(RzCore *core, GHT tls_addr, GHT tid) {
 		}
 
 		rz_io_nread_at(core->io, tls_addr, tcache_addr, sizeof(GHT));
-		GHT tcache_guess = GH(read_val)(core, tcache_addr, false);
+		const GHT tcache_guess = GH(read_val)(core, tcache_addr, false);
 		if (tcache_guess < map->addr || tcache_guess > map->addr_end) {
 			continue;
 		}
 
+#if __aarch64__
```
Comment (XVilka) [inline librz/core/linux_heap_glibc.c:864]: In the future, parsing should not depend on the architecture Rizin was built on.

**Review by XVilka — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Review by Rot127 — APPROVED** [review]

File: librz/bin/bin.c [hunk 5]
```diff
@@ -200,13 +203,8 @@ RZ_API void rz_bin_reloc_free(RZ_NULLABLE RzBinReloc *reloc) {
...
 	}
-	/**
-	 * TODO: leak in bin_elf, but it will cause double free in bin_pe if free here,
-	 * Because in the bin_elf implementation RzBinObject->imports and RzBinObject->relocs->imports
-	 * are two pieces of data, but they are linked to each other in bin_pe
-	 */
-	//	rz_bin_import_free(reloc->import);
-	//	rz_bin_symbol_free(reloc->symbol);
+	rz_bin_import_free(reloc->import);
+	rz_bin_symbol_free(reloc->symbol);
```
Comment (p371jah) [inline librz/bin/bin.c:207]: These changes are not related to heap parsing and not tested well. Reloc related changes cause UAF and must be reverted.

Comment (p371jah) [reply librz/bin/bin.c:207]: @XVilka

Comment (XVilka) [reply librz/bin/bin.c:207]: Ok. @giridharprasath could you please send a PR reverting those? @pelijah do you have an example of the test file where it happens?

Comment (giridharprasath) [reply librz/bin/bin.c:207]: Sure @XVilka, will send them. I have tested this with ELF and PE binaries. The fix caused UAF in mach binaries, I have fixed there as well. @pelijah Please add where it is reproducible.

Comment (p371jah) [reply librz/bin/bin.c:207]: COFF files and some LE libs

Comment (p371jah) [reply librz/bin/bin.c:207]: There was some reason why relocs refer to original imports/symbols. I have concerns that breaking this unwritten rule may cause side effects.
